specs/interop/token-bridging.md

```diff
@@ -2,96 +2,296 @@
+  return sendERC20To(msg.sender, _amount, _chainId);
+}
 
-  function bridgeERC20(uint256 _amount, uint256 _chainId) external {
-    bridgeERC20To(msg.sender, _amount, _chainId);
-  }
+function sendERC20To(address _to, uint256 _amount, uint256 _chainId) public returns (bool success) {
+  _burn(msg.sender, _amount);
+  bytes memory _message = abi.encodeWithSignature("finalizeSendERC20(address,uint256)", _to, _amount);
+  _sendMessage(_chainId, _message);
+  return true
+}
 
-  function bridgeERC20To(address _to, uint256 _amount, uint256 _chainId) public {
-    _burn(msg.sender, amount);
+function sendERC20To(address _to, uint256 _amount, uint256 _chainId, bytes memory _data) public returns (bool success) {
+  _burn(msg.sender, _amount);
+  bytes memory _message = abi.encodeWithSignature("finalizeSendERC20(address,uint256,bytes)", _to, _amount, _data);
+  _sendMessage(_chainId, _message);
+  return true
+}
 
-    L2ToL2CrossDomainMessenger.sendMessage({
-      _destination: chainId,
-      _target: address(this),
-      _message: abi.encodeCall(this.finalizeBridgeERC20, (to, amount))
-    });
-  }
+function finalizeSendERC20(address _to, uint256 _amount) external {
+  // this will be a modifier, inline here for clarity
+  require(msg.sender == address(L2ToL2CrossChainMessenger));
+  require(L2ToL2CrossChainMessenger.crossDomainMessageSender() == address(this));
+  //
+  _mint(_to, _amount);
+}
 
-  function finalizeBridgeERC20(address _to, uint256 _amount) external {
-    require(msg.sender == address(L2ToL2CrossChainMessenger));
-    require(L2ToL2CrossChainMessenger.crossDomainMessageSender() == address(this));
+function finalizeSendERC20(address _to, uint256 _amount, bytes memory _data) external {
+  require(msg.sender == address(L2ToL2CrossChainMessenger));
+  require(L2ToL2CrossChainMessenger.crossDomainMessageSender() == address(this));
+  _mint(_to, _amount);
+  (bool success, ) = _to.call(_data);
+  require(success, "External call failed");
+}
 
-    _mint(to, amount);
-  }
+/////// Internal functions ///////////
+function _sendMessage(uint256 _destination, bytes memory _data) internal virtual {
+  L2ToL2CrossDomainMessenger.sendMessage(_destination, address(this), _data);
+}
+```
+
+Note: Some other naming options that were considered were `bridgeERC20`, `send()`, `xTransfer()`, `Transfer` (with a chainId parameter).
+
+## Invariants
+
+Besides the ERC20 invariants, the SuperchainERC20 will require the following interop specific properties:
+
+- Conservation of finalized `totalSupply`: The minted `amount` in `finalizeSendERC20()` should match the `amount` that was burnt in `sendERC20To()`, as long as target chain has the initiating chain in the dependency set. This implies that finalized cross-chain transactions will conserve the sum of `totalSupply` for each chain in the Superchain.
+  - Corollary 1: each in flight messages will decrease the `totalSupply` exactly by the burnt `amount`.
```

Comment:
Does "in flight" mean "initiated message is safe but executing message has not happened yet"? Is it possible the executing message never occurs?

I would redefine this invariant so there is no corollary to simplify it, e.g. the "total supply is conserved across all safe blocks" (or whatever the actual invariant is)

Reply:
yes, in flight means exactly that, I can clarify. Both sentences are equivalent I think, so we can go with whatever you consider more clear.

Reply:
As we talked about during the design my example suggestion of `"total supply is conserved across all safe blocks" (or whatever the actual invariant is)` is not practical, since there is no guarantee an executing message is ever sent.

From my new understanding, some invariants we have are:
- For a fixed supply token, the sum of totalSupply across chains should never increase. (It stays the same when every initiating message has its executing message, but decreases if some executing messages are never sent)
    - This expands similarly for a token that has some way to be minted:
        - In the absence of non-crosschain-mints (i.e. ones that are expected to increase total supply) we get the fixed supply token invariant
        - When there are non-crosschain-mints, the sum of total supply across chains increases by the exact non-crosschain-mint amount
- The sum of all user balances across all chains must be exactly equal to the sum of the total supply across all chains. 
    - We preserve the regular ERC20 invariant that the sum of all user balances on a single chain must be exactly equal to total supply on that chain

Reply:
I like the fixed supply clarification. 
Also, the sum of user's balance vs total supply is a very strong invariant we had missed, good catch.

Adding on top of that one, the sum of each user's balance across chains will behave like the totalSupply invariant, i.e. it should not increase for tokens of fix supply.

I do feel like "sum of totalSupply across chains should never increase" is not precise enough. In any situation where this is not conserved (for fixed supply tokens) then its decrease exactly the bridged amount and it's because the transaction has not been completed.

Reply:
Regarding the cross-chain mint, do you feel like this is something that should be implemented inside the sendERC20/relayERC20 logic?